libs/vkd3d/state.c

```diff
@@ -4186,18 +4186,170 @@ static void d3d12_pipeline_state_graphics_handle_meta(struct d3d12_pipeline_stat
+        {
+            /* For tessellation shaders, the component count must match exactly. */
+            mismatch = mismatch || ((in_element->mask & 0xf) != (out_element->mask & 0xf));
+        }
+        else
+        {
+            /* Otherwise, it is legal to consume only a subset of components provided by the
+             * previous stage, but consuming components not provided by the previous stage is not. */
+            mismatch = mismatch || !!(in_element->mask & ~out_element->mask & 0xf);
+        }
+
+        if (mismatch)
+        {
+            WARN("Input signature element %s%u (reg %u, mask %#x) not compatible with %s%u (reg %u, mask %#x).\n",
+                    in_element->semantic_name, in_element->semantic_index, in_element->register_index, in_element->mask & 0xf,
+                    out_element->semantic_name, out_element->semantic_index, out_element->register_index, out_element->mask & 0xf);
+            return false;
+        }
+    }
+
+    /* The domain shader must consume all hull shader outputs, including
+     * tessellation factors. */
+    if (input_stage == VK_SHADER_STAGE_TESSELLATION_EVALUATION_BIT)
+    {
+        for (i = 0; i < out_sig->element_count; i++)
+        {
+            out_element = &out_sig->elements[i];
+
+            if (!(vkd3d_shader_find_signature_element(in_sig, out_element->semantic_name,
+                    out_element->semantic_index, out_element->stream_index)))
+            {
+                WARN("Hull shader output %s%u not consumed by domain shader.\n",
+                        out_element->semantic_name, out_element->semantic_index);
+                return false;
+            }
+        }
+    }
+
+    return true;
+}
+
+static bool vkd3d_validate_mesh_shader_io_signatures(const struct vkd3d_shader_signature *vert_sig,
+        const struct vkd3d_shader_signature *prim_sig, const struct vkd3d_shader_signature *in_sig)
+{
+    const struct vkd3d_shader_signature_element *in_element, *out_element;
+    unsigned int i;
+    bool mismatch;
+
+    for (i = 0; i < in_sig->element_count; i++)
+    {
+        in_element = &in_sig->elements[i];
+
+        if (!(out_element = vkd3d_shader_find_signature_element(vert_sig, in_element->semantic_name, in_element->semantic_index, 0)) &&
+                !(out_element = vkd3d_shader_find_signature_element(prim_sig, in_element->semantic_name, in_element->semantic_index, 0)))
+        {
+            if (vkd3d_shader_semantic_is_generated_for_stage(in_element->sysval_semantic, VK_SHADER_STAGE_MESH_BIT_EXT, VK_SHADER_STAGE_FRAGMENT_BIT))
+                continue;
+
+            WARN("No corresponding output signature element found for %s%u.\n",
+                    in_element->semantic_name, in_element->semantic_index);
+            return false;
+        }
+
+        mismatch = in_element->component_type != out_element->component_type ||
+                vkd3d_popcount(in_element->mask & 0xf) != vkd3d_popcount(out_element->mask & 0xf) ||
```

Comment:
Is it not allowed to consume partial vector in MS -> PS?

Reply:
Correct.